### PR TITLE
(595) Set up HMPPS Audit for Refer

### PR DESCRIPTION
**Pull request by Gweaton**
## Context

https://trello.com/c/PfPffKbd/595-set-up-hmpps-audit-for-refer

As per [this ADR](https://dsdmoj.atlassian.net/wiki/spaces/NDSS/pages/4061822977/ADR+-+Audit+event+capture), there's a requirement for services at HMPPS to send audit events of:

- All updates to data
- All views of personal data

And the following details:

- User identifier - if in doubt, username or email address
- Date and time - to at least the second level, ideally to millisecond
- Data entity identifier - usually the most sensible ID for the Person record being viewed - CRN, Prison Number or something else
- What is being viewed - not too fine-grained - “person risks” is appropriate, a list of each of the risks would not be. We think storing the page name might work, and if there’s a tab structure within pages, storing the tab name may also be appropriate. It is deliberate that we are talking about front-end interactions here - “what was the user doing?” rather than “what was the system doing to support the user’s request?”.

## Changes in this PR

This adds auditing events to the routes in the Refer journey, sending them to HMPPS Audit alongside any important information like the prisonNumber of the person being searched for. This is handled by new audit middleware.

I've followed the approach taken by the CAS teams [here](https://github.com/ministryofjustice/hmpps-approved-premises-ui/pull/886) - the code is fairly complex and could possibly simplified but in the interests of time I've kept things fairly as-is.

We need to set up some infrastructure on the Cloud Platform environment here to allow this to work.

### Examples of auditing logged in the terminal

image

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the HMPPS Cloud Platform environments repository for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/middleware/auditMiddleware.ts [hunk 1]
```diff
@@ -3,8 +3,27 @@ import type { NextFunction, Request, RequestHandler, Response } from 'express'
...
+  }
+}
+
 const wrapHandler =
-  (handler: RequestHandler, auditService: AuditService, auditEvent: string | undefined) =>
+  (
+    handler: RequestHandler,
+    auditService: AuditService,
+    auditEvent: string | undefined,
+    auditBodyParams: Array<string> | undefined,
```
Comment (yndajas) [inline server/middleware/auditMiddleware.ts:45]: Could these be `?` params?

File: server/middleware/auditMiddleware.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import type { NextFunction, Request, RequestHandler, Response } from 'express'
+import { type Key, pathToRegexp } from 'path-to-regexp'
```
Comment (yndajas) [inline server/middleware/auditMiddleware.ts:2]: This is what I was talking about the other day. This should really be two lines, one for the type and one for the non-type import, to fit our style. Should probably work out if we can get ESLint to enforce this (again?)

File: server/middleware/auditMiddleware.ts [hunk 3]
```diff
@@ -17,14 +18,37 @@ const auditDetails = (req: Request, auditBodyParams: Array<string> | undefined)
...
 
+const matchAuditEvent = (
+  path: string,
+  redirectMatcher: RedirectAuditMatcher,
+  redirectParams: Record<string, string>,
+) => {
+  const matches = redirectMatcher.regExp.exec(path)
+
+  if (matches) {
+    redirectMatcher.keys.forEach((key, i) => {
```
Comment (yndajas) [inline server/middleware/auditMiddleware.ts:29]: `keyIndex`?

File: server/middleware/auditMiddleware.ts [hunk 4]
```diff
@@ -34,18 +58,52 @@ const wrapHandler =
 
     await handler(req, res, next)
 
+    const encodedRedirectLocation = res.get('Location')
+    if (encodedRedirectLocation && redirectMatchers) {
+      const redirectPath = decodeURI(encodedRedirectLocation)
+      redirectParams = {}
```
Comment (yndajas) [inline server/middleware/auditMiddleware.ts:64]: I think this is setting `redirectParams` to what it already is?

File: server/middleware/auditMiddleware.test.ts [hunk 5]
```diff
@@ -125,4 +126,103 @@ describe('auditMiddleware', () => {
       })
     })
   })
+
+  describe('redirectAuditSpecs', () => {
+    const programmeHistoryPathPrefix = path('/referrals').path(':referralId').path('programme-history')
```
Comment (yndajas) [inline server/middleware/auditMiddleware.test.ts:131]: Can this now be imported from our predefined paths?

File: server/middleware/auditMiddleware.test.ts [hunk 6]
```diff
@@ -125,4 +126,103 @@ describe('auditMiddleware', () => {
...
+      expect(handler).toHaveBeenCalled()
+      expect(auditService.sendAuditMessage).toHaveBeenCalledWith('MATCHING_PATH_1_AUDIT_EVENT', username, {
+        referralId: 'my-referral-uuid',
+      })
+    })
+
+    it('sends the default message if the path does not match', async () => {
+      const nonMatchingPath = path('/').path('some-path-component')
+      const matchingPath1 = programmeHistoryPathPrefix.path('new')
+      const matchingPath2 = programmeHistoryPathPrefix.path(':courseParticipationId')
```
Comment (yndajas) [inline server/middleware/auditMiddleware.test.ts:199]: Having matching paths in a test about there not being a match is a little confusing

File: integration_tests/tsconfig.json [hunk 7]
```diff
@@ -3,6 +3,7 @@
     "strict": true,
     "target": "es5",
     "noEmit": true,
+    "skipLibCheck": true,
```
Comment (yndajas) [inline integration_tests/tsconfig.json:6]: Do we need to target `es5` here? No idea if that would fix things but we target `es2018` in the main `tsconfig.json`

File: server/utils/routeUtils.ts [hunk 8]
```diff
@@ -3,20 +3,63 @@
 import type { RequestHandler, Router } from 'express'
 
 import asyncMiddleware from '../middleware/asyncMiddleware'
+import type { AuditEventSpec } from '../middleware/auditMiddleware'
+import auditMiddleware from '../middleware/auditMiddleware'
```
Comment (yndajas) [inline server/utils/routeUtils.ts:7]: The `auditMiddleware` can be imported from the index. Could the type be exported/imported from the index too?

File: server/app.ts [hunk 9]
```diff
@@ -49,7 +49,7 @@ export default function createApp(controllers: Controllers, services: Services):
   app.use(setUpCsrf())
   app.use(setUpCurrentUser(services))
 
-  app.use(routes(controllers))
+  app.use(routes(controllers, services))
```
Comment (yndajas) [inline server/app.ts:52]: Do we need to pass the entire `services` object or just the audit service?

File: server/routes/refer.ts [hunk 10]
```diff
@@ -22,40 +22,133 @@ export default function routes(controllers: Controllers, services: Services, rou
...
+    auditEvent: 'REFER_VIEW_REFERRED_PERSON',
+  })
 
-  get(referPaths.confirmOasys.show.pattern, oasysConfirmationController.show())
-  put(referPaths.confirmOasys.update.pattern, oasysConfirmationController.update())
+  get(referPaths.confirmOasys.show.pattern, oasysConfirmationController.show(), {
+    auditEvent: 'REFER_VIEW_CONFIRM_OASYS',
+  })
+  put(referPaths.confirmOasys.update.pattern, oasysConfirmationController.update(), {
+    redirectAuditEventSpecs: [
```
Comment (yndajas) [inline server/routes/refer.ts:47]: I get what this is doing but `redirectAuditEventSpecs` is a bit of a brain melting combination of words 😅

Also, will we get a test failure if we add a new redirect or it redirects somewhere unexpected? I can see us forgetting to update this if adding/changing redirects in existing pages,

**Review by yndajas — COMMENTED** [review]
This is complicated 🤯 Would definitely benefit from simplification. I gave up trying to make sense of the middleware on the fifth commit - some of what I suggested in earlier commits might not be so relevant as it gets more and more complicated

I also wonder how well protected from we are here from mistakes when setting up the events, or just forgetting to set them up altogether? I think there are a few mistakes in the Refer routes file that aren't captured by tests?

**Comment by jsrobertson** [comment]
Closing for now until we're ready and in a position to pick this up again.

**Pull request closed without merging**